File: docs/src/main/scala/sexamples/simulation/pingpongglobal/Pinger.scala
package sexamples.simulation.pingpongglobal

import se.sics.kompics.sl._
import se.sics.kompics.network.Network
import se.sics.kompics.timer.{CancelPeriodicTimeout, SchedulePeriodicTimeout, Timeout, Timer}
import se.sics.kompics.simulator.util.GlobalView

import java.util.UUID

object Pinger {
  class PingTimeout(_spt: SchedulePeriodicTimeout) extends Timeout(_spt);
}
class Pinger extends ComponentDefinition {
  import Pinger.PingTimeout;

  val self = cfg.getValue[TAddress]("pingpong.pinger.addr");
  val ponger = cfg.getValue[TAddress]("pingpong.pinger.pongeraddr");
  val timeoutPeriod = cfg.getValue[Long]("pingpong.pinger.timeout");

  val net = requires[Network];
  val timer = requires[Timer];

  private var counter: Long = 0L;
  private var timerId: Option[UUID] = None;

  ctrl uponEvent {
    case _: Start => {
      val spt = new SchedulePeriodicTimeout(0, timeoutPeriod);
      val timeout = new PingTimeout(spt);
      spt.setTimeoutEvent(timeout);
      trigger(spt -> timer);
      timerId = Some(timeout.getTimeoutId());
    }
  }

  // #ponged
  net uponEvent {
    case TMessage(_, Pong) => {
      counter += 1L;
      log.info(s"Got Pong #${counter}!");
      ponged()
    }
  }
  // #ponged

  timer uponEvent {
    case _: PingTimeout => {
      trigger(TMessage(self, ponger, Ping) -> net);
    }
  }

  override def tearDown(): Unit = {
    timerId match {
      case Some(id) => {
        trigger(new CancelPeriodicTimeout(id) -> timer);
      }
      case None => () // no cleanup necessary
    }
  }

  // #ponged
  private def ponged(): Unit = {
    val gvo: Option[GlobalView] = cfg.readValue[GlobalView]("simulation.globalview");
    gvo match {
      case Some(gv: GlobalView) => {
        // Java API only :(
        gv.setValue("simulation.pongs", gv.getValue("simulation.pongs", classOf[Integer]) + 1);
      }
      case None => throw new RuntimeException("Why are we not in simulation?") // not in simulation
    }
  }
  // #ponged

}

File: docs/src/main/java/jexamples/simulation/pingpongglobal/Pinger.java
package jexamples.simulation.pingpongglobal;

import se.sics.kompics.Kompics;
import se.sics.kompics.ClassMatchedHandler;
import se.sics.kompics.ComponentDefinition;
import se.sics.kompics.Positive;
import se.sics.kompics.Handler;
import se.sics.kompics.Start;
import se.sics.kompics.timer.*;
import se.sics.kompics.network.Network;
import se.sics.kompics.network.Transport;
import se.sics.kompics.simulator.util.GlobalView;

import java.util.UUID;
import java.util.Optional;

public class Pinger extends ComponentDefinition {

  Positive<Network> net = requires(Network.class);
  Positive<Timer> timer = requires(Timer.class);

  private long counter = 0;
  private UUID timerId;
  private final TAddress self;
  private final TAddress ponger;
  private final long timeoutPeriod;

  public Pinger() {
    this.self = config().getValue("pingpong.pinger.addr", TAddress.class);
    this.ponger = config().getValue("pingpong.pinger.pongeraddr", TAddress.class);
    this.timeoutPeriod = config().getValue("pingpong.pinger.timeout", Long.class);
  }

  Handler<Start> startHandler =
      new Handler<Start>() {
        public void handle(Start event) {
          SchedulePeriodicTimeout spt = new SchedulePeriodicTimeout(0, timeoutPeriod);
          PingTimeout timeout = new PingTimeout(spt);
          spt.setTimeoutEvent(timeout);
          trigger(spt, timer);
          timerId = timeout.getTimeoutId();
        }
      };
  // #ponged
  ClassMatchedHandler<Pong, TMessage> pongHandler =
      new ClassMatchedHandler<Pong, TMessage>() {

        @Override
        public void handle(Pong content, TMessage context) {
          counter++;
          logger.info("Got Pong #{}!", counter);
          ponged();
        }
      };
  // #ponged
  Handler<PingTimeout> timeoutHandler =
      new Handler<PingTimeout>() {
        public void handle(PingTimeout event) {
          trigger(new TMessage(self, ponger, Transport.TCP, Ping.EVENT), net);
        }
      };

  {
    subscribe(startHandler, control);
    subscribe(pongHandler, net);
    subscribe(timeoutHandler, timer);
  }

  @Override
  public void tearDown() {
    trigger(new CancelPeriodicTimeout(timerId), timer);
  }
  // #ponged
  private void ponged() {
    Optional<GlobalView> gvo = config().readValue("simulation.globalview", GlobalView.class);
    if (gvo.isPresent()) { // only in simulation
      GlobalView gv = gvo.get();
      gv.setValue("simulation.pongs", gv.getValue("simulation.pongs", Integer.class) + 1);
    }
  }
  // #ponged

  public static class PingTimeout extends Timeout {
    public PingTimeout(SchedulePeriodicTimeout spt) {
      super(spt);
    }
  }
}
